Google Sheets connector

puneetjaiswal opened the pull request:
Instructions - https://docs.google.com/spreadsheets/d/1Es4HhWALUQjoa-bQh4a8B5HROz7dpGMfq_HbfoaW5LM/edit#gid=353045147

Metadata sheet - https://docs.google.com/spreadsheets/d/1Es4HhWALUQjoa-bQh4a8B5HROz7dpGMfq_HbfoaW5LM/edit#gid=0

Test table - https://docs.google.com/spreadsheets/d/1S625j2oTptRepg6Yci68fCYE1269tdoSjljNOmTgQ3U/edit#gid=0

kokosing commented on pom.xml:
```diff
@@ -135,6 +135,7 @@
         <module>presto-proxy</module>
         <module>presto-kudu</module>
         <module>presto-elasticsearch</module>
+        <module>presto-sheets</module>
```
presto-google-sheets?

kokosing commented on presto-sheets/pom.xml:
```diff
@@ -0,0 +1,170 @@
+    </parent>
+
+    <artifactId>presto-sheets</artifactId>
+    <description>Presto - Sheets Readonly Connector</description>
```
Presto - Google Sheets Connector?

I do not think there is a point to emphasise that it is read only.

kokosing commented on presto-main/etc/catalog/sheets.properties:
```diff
@@ -0,0 +1,9 @@
+connector.name=sheets
+credentials-path={{ service account credentials file path }}
```
maybe hardcode some arbitrary path that deloverper can have locally. Also you can mention in Readme how such file can be generated (if you haven't already).

kokosing commented on presto-sheets/pom.xml:
```diff
@@ -0,0 +1,170 @@
+            <artifactId>jackson-databind</artifactId>
+        </dependency>
+
+        <!-- gsheets deps begins -->
```
I would drop `begins`

kokosing commented on presto-sheets/pom.xml:
```diff
@@ -0,0 +1,170 @@
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <!-- gsheets deps ends  -->
```
I would drop this comment.

kokosing commented on presto-sheets/src/main/java/io/prestosql/plugin/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,85 @@
+        catch (Exception e) {
+            log.error(e, "Error accessing table info [%s]", tableName);
+        }
+        return null;
```
return Optional

kokosing commented on presto-sheets/src/main/java/io/prestosql/plugin/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,85 @@
+            }
+        }
+        catch (Exception e) {
+            log.error(e, "Error accessing table info [%s]", tableName);
```
why not to simply fail?

kokosing commented on presto-sheets/src/main/java/io/prestosql/plugin/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,85 @@
+    {
+        requireNonNull(config, "config is null");
+        requireNonNull(catalogCodec, "catalogCodec is null");
+        this.schema = new HashSet<>();
```
ImmutableSet.of("sheets")?

findepi replied:
When we don't have schema names, we often use `default` for the only schema's name.

kokosing commented on presto-sheets/src/main/java/io/prestosql/plugin/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,85 @@
+
+public class SheetsClient
+{
+    private Set<String> schema;
```
final

kokosing commented on presto-sheets/src/test/java/io/prestosql/plugin/sheets/SheetsSimpleIntegrationTest.java:
```diff
@@ -0,0 +1,92 @@
+            SheetsPlugin sheetsPlugin = new SheetsPlugin();
+            queryRunner = DistributedQueryRunner.builder(createSession()).build();
+            queryRunner.installPlugin(sheetsPlugin);
+            queryRunner.createCatalog(SHEETS, SHEETS, ImmutableMap.of("credentials-path", getTestCredentialsPath(),
```
please put each arg into seperate line, like

```
method(
  arg1,
  arg2,
  arg3);
```

kokosing commented on presto-sheets/src/test/resources/gsheets-creds-base64-encoded.props:
```diff
@@ -0,0 +1 @@
+ewogICJ0eXBlIjogInNlcnZpY2VfYWNjb3VudCIsCiAgInByb2plY3RfaWQiOiAicHJlc3RvdGVzdCIsCiAgInByaXZhdGVfa2V5X2lkIjogIjIxMDFiMzBkNjM2ZWU3ZWNkZmFjN2U3M2JiOThmOTVjNjY5OTE5ZmQiLAogICJwcml2YXRlX2tleSI6ICItLS0tLUJFR0lOIFBSSVZBVEUgS0VZLS0tLS1cbk1JSUV1d0lCQURBTkJna3Foa2lHOXcwQkFRRUZBQVNDQktVd2dnU2hBZ0VBQW9JQkFRQ3RjdGl3YmtpZ0s0N29cbjR3VGFJU0lNcEFaTXV6V0h4emF4andReUYzZk93QUZSWlk0Vm8yOTlkWi9zeDcvdU94Nk5EMmRIbDFZa2J ...
```
are there any quota on this? If we start to use travis on for each PR we might hit some google limits here.

findepi replied:
https://github.com/prestosql/presto/pull/973 allows us to run certain tests on master, but not on PR.

btw what are these credentials to? What account?

puneetjaiswal replied:
Api docs says unlimited quota, but there should be a limit of 300 reqs per sec.
I created a test google account - `prestosqltest@gmail.com`. Please let me know how I can share the login info etc, or if there is another test account.

findepi replied:
> I created a test google account - `prestosqltest@gmail.com`. Please let me know how I can share the login info etc, or if there is another test account.

cc @martint @electrum

dain replied:
We go with this account, and when it is checked in, we can switch to an account at prestosql.io, which is in gsuite.

kokosing commented on presto-sheets/src/test/java/io/prestosql/plugin/sheets/SheetsSimpleIntegrationTest.java:
```diff
@@ -0,0 +1,92 @@
+import static io.prestosql.plugin.sheets.TestSheetsPlugin.getTestCredentialsPath;
+import static io.prestosql.testing.TestingSession.testSessionBuilder;
+
+public class SheetsSimpleIntegrationTest
```
Please rename to `TestGoogleSheets`. Convention in Presto is to prefix test class names from `Test`

kokosing reviewed (commented):
Very nice piece of code. Just skimmed.

puneetjaiswal commented:
@findepi / @kokosing - thanks for reviewing :) 
I have updated the PR, addressing the comments. IMO integ tests would not cause failures with google API rate limits, so keeping the tests to run with each PR. Please let me know if you would like this to be triggered with master build only, I would be happy to refactor, thanks!

kokosing commented:
Please correct me if I am wrong. As I understand you define a sheet with metadata first. The one which contains information about other tables (sheets). Why do we do this way? In other connectors we specify such metadata next to Presto configuration files as JSON (see kafka connector). We could also implement refreshing (like it is used for file based access control). Having sheet for metadata makes that configuration of Presto is distributed among different services and it could be difficult to maintain it. Also I think it is easier to leak an access to a google sheet than a Presto configuration file. What do you think?

puneetjaiswal commented:
> Please correct me if I am wrong. As I understand you define a sheet with metadata first. The one which contains information about other tables (sheets). Why do we do this way?

We want our users to be able to add, remove, rename tables on the fly and so we would need an online store for table metadata. Since we are already connecting to gsheets API for reading sheets data, I thought of using it for storing table metadata too through a metadata sheet.

> In other connectors we specify such metadata next to Presto configuration files as JSON (see kafka connector). We could also implement refreshing (like it is used for file based access control). Having sheet for metadata makes that configuration of Presto is distributed among different services and it could be difficult to maintain it. 

If we keep the table metadata in static config files, any addition, removal, renaming of tables would require config change and thus a deployment leading to slow/bad user experience. 

> Also I think it is easier to leak an access to a google sheet than a Presto configuration file. What do you think?

Currently its open for a selected internal users. My plans are to add a sheet - table onboarding UI where we can authenticate user and let them edit their table's metadata only, this way we would not have to share the metastore sheet with user. 

Lets connect on slack and discuss these in details.
Thanks again for reviewing, really appreciate this :)

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,81 @@
+            }
+        }
+        catch (Exception e) {
+            throw new IllegalArgumentException("Error accessing table - " + tableName, e);
```
PrestoException with some connector related error

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsConfig.java:
```diff
@@ -0,0 +1,74 @@
+        return this;
+    }
+
+    public String getMetadataSheetId()
```
Optional? or `@NotNull`

puneetjaiswal replied:
added @NotNull

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsConfig.java:
```diff
@@ -0,0 +1,74 @@
+    private int sheetsDataExpireAfterAccessMinutes = 1;
+    private int sheetsDataExpireAfterWriteMinutes = 5;
+
+    public String getCredentialsFilePath()
```
Optional? or `@NotNull`

puneetjaiswal replied:
added @NotNull

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsMetadata.java:
```diff
@@ -0,0 +1,154 @@
+    private ConnectorTableMetadata getTableMetadata(SchemaTableName tableName)
+    {
+        if (!listSchemaNames().contains(tableName.getSchemaName())) {
+            return null;
```
`return Optional.empty()`?

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsRecordCursor.java:
```diff
@@ -0,0 +1,163 @@
+
+    public SheetsRecordCursor(List<SheetsColumnHandle> columnHandles, List<List<Object>> dataValues)
+    {
+        this.columnHandles = columnHandles;
```
rnn

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsRecordCursor.java:
```diff
@@ -0,0 +1,163 @@
+    public SheetsRecordCursor(List<SheetsColumnHandle> columnHandles, List<List<Object>> dataValues)
+    {
+        this.columnHandles = columnHandles;
+        this.dataValues = dataValues;
```
rnn and ImmutableList.copyOf

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsRecordCursor.java:
```diff
@@ -0,0 +1,163 @@
+        // Populate incomplete columns with null
+        String[] allFields = new String[columnHandles.size()];
+
+        if (currentVals.size() > 0) {
```
`currentVals.size() == 0` should never happen according to above while loop condition, you can replace this with checkstate

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsRecordSet.java:
```diff
@@ -0,0 +1,55 @@
+        this.values = split.getValues();
+        ImmutableList.Builder<Type> types = ImmutableList.builder();
+        for (SheetsColumnHandle column : columnHandles) {
+            types.add(column.getColumnType());
```
columnHandles.stream().map(...).collect(toImmutableList())

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsRecordSetProvider.java:
```diff
@@ -0,0 +1,44 @@
+
+        ImmutableList.Builder<SheetsColumnHandle> handles = ImmutableList.builder();
+        for (ColumnHandle handle : columns) {
+            handles.add((SheetsColumnHandle) handle);
```
stream() and collect(toImmutableList())

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/provider/SheetsDataProvider.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.List;
+import java.util.Set;
+
+public abstract class SheetsDataProvider
```
why do you need this abstract class?

puneetjaiswal replied:
We dont, I will remove this.

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/provider/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,181 @@
+            this.sheetsService = new Sheets.Builder(httpTransport, JSON_FACTORY, this.credential).setApplicationName(APPLICATION_NAME).build();
+        }
+        catch (Exception e) {
+            log.error(e, "Unable to load metadata mapping from %s", config.getMetadataSheetId());
```
shouldn't we fail initialization here?

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/provider/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,181 @@
+                {
+                    public String load(String tableName)
+                    {
+                        try {
```
extract this block as method. In general please follow the convention like in `io.prestosql.plugin.hive.metastore.CachingHiveMetastore`

kokosing commented on presto-google-sheets/src/test/java/io/prestosql/plugin/google/sheets/TestSheetsPlugin.java:
```diff
@@ -0,0 +1,72 @@
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static org.testng.Assert.assertNotNull;
+
+public class TestSheetsPlugin
```
can you please prepare TPCH data in google sheets format so we can run regular test suite.

puneetjaiswal replied:
Since this connector only supports `varchar` column type, we may not be able to run regular tests yet. Once we add better support to column type mapping, we can add the TPCH data based tests.

kokosing reviewed (commented):
In general looks good to me. Just few no big deal comments.

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,205 @@
+    private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+
+    private static final List<String> SCOPES =
+            Collections.singletonList(SheetsScopes.SPREADSHEETS_READONLY);
```
no point to break the line here. Also Presto codebase prefers using `ImmutableList.of(SheetsScopes.SPREADSHEETS_READONLY)`

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,205 @@
+
+    public GSheetsDataProvider(SheetsConfig config)
+    {
+        this.sheetsConfig = config;
```
we typically do not store entire config as fields, but take only required parts out of it. For example please see `io.prestosql.plugin.hive.HiveSplitManager#HiveSplitManager`

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,205 @@
+        }
+        OptionalLong expiresAfterWriteMillis = OptionalLong.of(TimeUnit.MINUTES.toMillis(sheetsConfig.getSheetsDataExpireAfterWriteMinutes()));
+        OptionalLong refreshMillis = OptionalLong.of(TimeUnit.MINUTES.toMillis(sheetsConfig.getSheetsDataExpireAfterAccessMinutes()));
+        long maxCacheSize = 1000;
```
shouldn't this be in the config?

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,205 @@
+        OptionalLong expiresAfterWriteMillis = OptionalLong.of(TimeUnit.MINUTES.toMillis(sheetsConfig.getSheetsDataExpireAfterWriteMinutes()));
+        OptionalLong refreshMillis = OptionalLong.of(TimeUnit.MINUTES.toMillis(sheetsConfig.getSheetsDataExpireAfterAccessMinutes()));
+        long maxCacheSize = 1000;
+        ExecutorService executor = newDirectExecutorService();
```
does refreshing work with direct executor? Maybe let's remove the refreshing option.

puneetjaiswal replied:
good point, removing refreshing options.

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,205 @@
+    /**
+     * @return
+     * @throws IOException
+     */
```
please remove such javadocs

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,205 @@
+        return sheetDataCache.get(sheetId).get();
+    }
+
+    private List<List<Object>> readAllValuesFromSheetExpression(String sheetExpression) throws Exception
```
Do you think that introducing some entity for `List<List<Object>>` be beneficial? Like `Rows` that store `Row`.

puneetjaiswal replied:
Yes, I will add SheetRow / SheetRows once add better column type mapping, may be in next iteration if thats ok.

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,205 @@
+        log.info("Accessing sheet id [%s] with range [%s]", sheetId, defaultRange);
+        ValueRange response = valuesAll.get(sheetId, defaultRange).execute();
+        List<List<Object>> values = response.getValues();
+        return values;
```
inline `values`

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,205 @@
+        Sheets.Spreadsheets.Values valuesAll = sheetsService.spreadsheets().values();
+        log.info("Accessing sheet id [%s] with range [%s]", sheetId, defaultRange);
+        ValueRange response = valuesAll.get(sheetId, defaultRange).execute();
+        List<List<Object>> values = response.getValues();
```
inline `response`

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,82 @@
+    {
+        requireNonNull(config, "config is null");
+        requireNonNull(catalogCodec, "catalogCodec is null");
+        this.schema = ImmutableSet.of("default");
```
move this to line 35. This also could be static

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,82 @@
+
+public class SheetsClient
+{
+    private final Set<String> schema;
```
schemas?

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,82 @@
+
+    public Set<String> getSchemaNames()
+    {
+        return schema;
```
maybe inline `schemas`?

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,82 @@
+        this.sheetsDataProvider = SheetsDataProvider.getSheetsDataProvider(config);
+    }
+
+    public Set<String> getSchemaNames()
```
does it belong here? Maybe it should be moved to `ConnectorMetadata`?

puneetjaiswal replied:
ConnectorMetadata delegates to `sheetsClient.getSchemaNames()`
We will use this when we add multiple schema support in this connector,

kokosing replied:
So maybe let's add it then. Future is unknown things may change before you add multiple schema support.

puneetjaiswal replied:
Sounds good, I will clear this for now.

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsConfig.java:
```diff
@@ -0,0 +1,78 @@
+    private String metadataSheetId;
+
+    private int sheetsDataExpireAfterAccessMinutes = 5;
+    private int sheetsDataExpireAfterWriteMinutes = 5;
```
use `Duration`

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsDataProvider.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.List;
+import java.util.Set;
+
+public interface SheetsDataProvider
```
do you need this interface at all?

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsErrorCode.java:
```diff
@@ -0,0 +1,42 @@
+
+    SheetsErrorCode(int code, ErrorType type)
+    {
+        errorCode = new ErrorCode(code + 0x0100_0000, name(), type);
```
please use uniq prefix, see that atop is using `0x0500_0000`

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,227 @@
+                }
+            }
+        }
+        catch (Exception e) {
```
Throwing and catching plain `Exception` is too broad. Use more dedicated and more meaningful exceptions.

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,227 @@
+        }
+    }
+
+    private Credential getCredentials() throws IOException
```
this method does not throw anything

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,227 @@
+            }
+        }
+        catch (Exception e) {
+            log.error(e, "Error retrieving table metadata mapping");
```
please do not swallow exceptions

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,227 @@
+        Set<String> tables = new HashSet<>();
+        try {
+            List<List<Object>> tableMetadata = sheetDataCache.get(metadataSheetId).get();
+            if (tableMetadata.size() > 1) { // we need both table name and sheet id
```
if condition not needed

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,75 @@
+        requireNonNull(config, "config is null");
+        requireNonNull(catalogCodec, "catalogCodec is null");
+
+        this.sheetsDataProvider = new GSheetsDataProvider(config);
```
why not to `@Inject` `GSheetsDataProvider` directly (instead of `config`)

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsConfig.java:
```diff
@@ -0,0 +1,81 @@
+        return credentialsFilePath;
+    }
+
+    @Config("credentials-path")
```
Please also use `@ConfigDescription` here and everywhere else in config classes

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsConnectorFactory.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    @Override
+    public Connector create(String catalogName, Map<String, String> requiredConfig, ConnectorContext context)
```
s/requiredConfig/config

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsConnectorFactory.java:
```diff
@@ -0,0 +1,67 @@
+    {
+        requireNonNull(requiredConfig, "requiredConfig is null");
+        try {
+            // A plugin is not required to use Guice; it is just very convenient
```
remove this comment

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsMetadata.java:
```diff
@@ -0,0 +1,155 @@
+        }
+
+        SheetsTable table = sheetsClient.getTable(tableName.getSchemaName(), tableName.getTableName()).get();
+        if (table == null) {
```
it would be nice to use `Optional` instead of nullable value

puneetjaiswal commented:
Thanks @kokosing for reviewing the PR updates so patiently. I have updated the PR addressing all your comments covering:
1. catching and throwing specific exceptions.
2. replacing manual null checks with `!Optional.isPresent()`
Please take a look, thanks!

puneetjaiswal commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsErrorCode.java:
```diff
@@ -0,0 +1,42 @@
+    @Override
+    public ErrorCode toErrorCode()
+    {
+        return toErrorCode();
```
just noticed this bug.

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,211 @@
+        return tableSheetMap.build();
+    }
+
+    private Optional<List<List<Object>>> getDataForSheetExpression(String sheetExpression)
```
this method does not return `Optional.empty()` so it `Optional` is not needed here

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,211 @@
+    {
+        ImmutableMap.Builder<String, Optional<String>> tableSheetMap = ImmutableMap.builder();
+        try {
+            List<List<Object>> data = readAllValuesFromSheetExpression(metadataSheetId);
```
use `getDataForSheetExpression`, then you can remove try-catch

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,211 @@
+    {
+        GoogleCredential credential;
+        try (InputStream in = new FileInputStream(credentialsFilePath)) {
+            credential = GoogleCredential.fromStream(in).createScoped(SCOPES);
```
return here

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,211 @@
+
+    public Set<String> listTables()
+    {
+        Set<String> tables = new HashSet<>();
```
immutable set

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,211 @@
+        CacheBuilder<Object, Object> cacheBuilder = CacheBuilder.newBuilder();
+        cacheBuilder = cacheBuilder.expireAfterWrite(expiresAfterWriteMillis, MILLISECONDS);
+        cacheBuilder = cacheBuilder.maximumSize(maximumSize);
+        return cacheBuilder;
```
inline `cacheBuilder`

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java:
```diff
@@ -0,0 +1,211 @@
+        if (!sheetExpression.isPresent()) {
+            throw new PrestoException(SHEETS_UNKNOWN_TABLE_ERROR, "Data not found for table " + tableName);
+        }
+        return sheetDataCache.get(sheetExpression.get());
```
handle `ExecutionException` here

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,79 @@
+        this.sheetsDataProvider = sheetsDataProvider;
+    }
+
+    public Set<String> getTableNames(String schema)
```
unused parameter

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,79 @@
+        return sheetsDataProvider.listTables();
+    }
+
+    public Optional<SheetsTable> getTable(String schema, String tableName)
```
unused parameter

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,79 @@
+                for (Object col : header) {
+                    String colVal = String.valueOf(col);
+                    // when empty or repeated column header, adding a placeholder column name
+                    if ("".equals(colVal) || columnNames.contains(colVal)) {
```
use `calVal.isEmpty()`

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,79 @@
+                // Assuming 1st line is always header
+                List<Object> header = values.get().get(0);
+                int count = 0;
+                for (Object col : header) {
```
col -> column, colValue -> columnValue

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,79 @@
+                List<Object> header = values.get().get(0);
+                int count = 0;
+                for (Object col : header) {
+                    String colVal = String.valueOf(col);
```
col.toString().toLowerCase(ENGLISH)

Can you please add a test for sheet with columns `a` and `A`

puneetjaiswal replied:
Added test data [here](https://docs.google.com/spreadsheets/d/1Es4HhWALUQjoa-bQh4a8B5HROz7dpGMfq_HbfoaW5LM/edit#gid=0)

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsClient.java:
```diff
@@ -0,0 +1,72 @@
+
+public class SheetsClient
+{
+    private final GSheetsDataProvider sheetsDataProvider;
```
I would inline that `GSheetsDataProvider` into `SheetsClient`. It is not clear to me why this was extracted as separate entity

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsColumnHandle.java:
```diff
@@ -0,0 +1,97 @@
+    @Override
+    public int hashCode()
+    {
+        return columnName.hashCode();
```
why type and ordinal fields are not used in hashCode/equals? Let's include them here. Alternatively you can remove hashCode/equals all together.

If you are going to store this in a map you can always map that by column name (string).

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsConfig.java:
```diff
@@ -0,0 +1,86 @@
+        return this;
+    }
+
+    public int getSheetsDataMaxCacheSize()
```
`@Min(1)`?

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsTable.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @JsonProperty
+    public String getName()
```
it looks like it is not used, if so please remove

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsTable.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @JsonProperty
+    public List<SheetsColumn> getColumns()
```
it looks like it is not used, if so please remove

kokosing commented on presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/SheetsTableHandle.java:
```diff
@@ -0,0 +1,84 @@
+        implements ConnectorTableHandle
+{
+    private final String schemaName;
+    private final String tableName;
```
use `SchemaTableName` instead the two fields above

kokosing commented on presto-google-sheets/src/test/java/io/prestosql/plugin/google/sheets/TestSheetsPlugin.java:
```diff
@@ -0,0 +1,73 @@
+        Connector c = factory.create(
+                            TestGoogleSheets.GOOGLE_SHEETS,
+                            ImmutableMap.of("credentials-path", getTestCredentialsPath(),
+                        "metadata-sheet-id", TEST_METADATA_SHEET_ID),
```
formatting is odd here. Please use map builder instead.

```
ImmutableMap.builder()
 .put()
 .put()
 .build()
```

kokosing commented on presto-google-sheets/src/test/java/io/prestosql/plugin/google/sheets/TestSheetsPlugin.java:
```diff
@@ -0,0 +1,73 @@
+            String credentials = new String(Base64.getDecoder().decode(credsBase64Encoded.toString()));
+
+            File tempFile =
+                    File.createTempFile(
```
there is no need to wrap the lines here

kokosing commented on presto-google-sheets/src/test/java/io/prestosql/plugin/google/sheets/TestSheetsPlugin.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Test
+    public void testCreateConnector() throws Exception
```
`throws` need to be in separate line

kokosing commented on presto-google-sheets/src/test/java/io/prestosql/plugin/google/sheets/TestSheetsPlugin.java:
```diff
@@ -0,0 +1,73 @@
+
+    static String getTestCredentialsPath() throws Exception
+    {
+        try (InputStream in =
```
no need to wrap line here

kokosing commented on presto-google-sheets/src/test/java/io/prestosql/plugin/google/sheets/TestSheetsConfig.java:
```diff
@@ -0,0 +1,55 @@
+    @Test
+    public void testDefaults()
+    {
+        ConfigAssertions.assertRecordedDefaults(
```
static imports, no need to wrap line here

kokosing reviewed (approved):
% comments

puneetjaiswal commented:
> % comments

Thanks for reviewing promptly. Resolved all the comments.

kokosing commented on presto-main/etc/catalog/gsheets.properties:
```diff
@@ -0,0 +1,20 @@
+############ How to generate google-sheets service account credentials ################
```
Instead of putting this here, let's create README.md file on top of `presto-google-sheets` which informations like that. I don't think this is easily findable. The next step would be to have proper documentation.

Commit:
Adding google sheets plugin

kokosing commented:
Merged, thanks!
